File: sdk/js-sdk/src/client/File.ts
/*
 * Copyright 2021 Layotto Authors
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { debuglog } from 'node:util';
import { Transform, Readable } from 'stream';
import { pipeline as pipelinePromise } from 'stream/promises';
import { 
  GetFileRequest as GetFileRequestPB,
  GetFileResponse as GetFileResponsePB,
  PutFileRequest as PutFileRequestPB,
  ListFileRequest as ListFileRequestPB,
  DelFileRequest as DelFileRequestPB,
} from '../../proto/runtime_pb';
import { API } from './API';
import { GetFileRequest, ListFileResponse, PutFileRequest } from '../types/File';

const debug = debuglog('layotto:client:file');

export default class File extends API {
  // Get a file stream
  async get(request: GetFileRequest): Promise<Readable> {
    const req = new GetFileRequestPB();
    req.setStoreName(request.storeName);
    req.setName(request.name);
    this.mergeMetadataToMap(req.getMetadataMap(), request.metadata);

    // convert GetFileResponsePB to Uint8Array
    const converter = new Transform({
      objectMode: true,
      transform(res: GetFileResponsePB, _, done) {
        const data = res.getData_asU8();
        done(null, data);
      }
    });
    const callStream = this.runtime.getFile(req, this.createMetadata(request));
    // Make sure callStream error handle by converter
    callStream.on('error', (err) => converter.emit('error', err));
    return callStream.pipe(converter);
  }

  async put(request: PutFileRequest): Promise<void> {
    const stream = request.stream ?? Readable.from(request.data);

    const ac = new AbortController();
    const signal = ac.signal;
    const writeStream = this.runtime.putFile(this.createMetadata(request), (err) => {
      if (err) {
        debug('putFile %j got server error: %s', request, err);
        // abort request and throw error
        // FIXME: should tell caller the real err reason, not only AbortError
        ac.abort();
      }
    });
    await pipelinePromise(
      stream,
      new Transform({
        objectMode: true,
        transform: (chunk: Uint8Array, _, done) => {
          const req = new PutFileRequestPB();
          req.setStoreName(request.storeName);
          req.setName(request.name);
          req.setData(chunk);
          this.mergeMetadataToMap(req.getMetadataMap(), request.metadata);
          done(null, req);
        }
      }),
      // @ts-ignore
      writeStream,
      { signal },
    );
  }

  async list(request: GetFileRequest): Promise<ListFileResponse> {
    const req = new GetFileRequestPB();
    req.setStoreName(request.storeName);
    req.setName(request.name);
    this.mergeMetadataToMap(req.getMetadataMap(), { storageType: 'Standard' }, request.metadata);
    const listReq = new ListFileRequestPB();
    listReq.setRequest(req);

    return new Promise((resolve, reject) => {
      this.runtime.listFile(listReq, this.createMetadata(request), (err, res) => {
        if (err) return reject(err);
        debug('listFile: %j, res: %j', request, res);
        resolve({
          names: res.getFileNameList(),
        });
      });
    });
  }

  async delete(request: GetFileRequest): Promise<void> {
    const req = new GetFileRequestPB();
    req.setStoreName(request.storeName);
    req.setName(request.name);
    const delReq = new DelFileRequestPB();
    delReq.setRequest(req);

    return new Promise((resolve, reject) => {
      this.runtime.delFile(delReq, this.createMetadata(request), (err) => {
        if (err) return reject(err);
        resolve();
      });
    });
  }
}

File: sdk/java-sdk/examples/src/test/java/io/mosn/layotto/examples/file/File.java
/*
 * Copyright 2021 Layotto Authors
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.mosn.layotto.examples.file;

import io.mosn.layotto.v1.RuntimeClientBuilder;
import io.mosn.layotto.v1.config.RuntimeProperties;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import spec.sdk.runtime.v1.client.RuntimeClient;
import spec.sdk.runtime.v1.domain.file.DelFileRequest;
import spec.sdk.runtime.v1.domain.file.FileInfo;
import spec.sdk.runtime.v1.domain.file.GetFileRequest;
import spec.sdk.runtime.v1.domain.file.GetFileResponse;
import spec.sdk.runtime.v1.domain.file.GetMetaRequest;
import spec.sdk.runtime.v1.domain.file.GetMeteResponse;
import spec.sdk.runtime.v1.domain.file.ListFileRequest;
import spec.sdk.runtime.v1.domain.file.ListFileResponse;
import spec.sdk.runtime.v1.domain.file.PutFileRequest;

import java.io.ByteArrayInputStream;
import java.io.InputStream;
import java.util.HashMap;
import java.util.Map;

/**
 * Specially
 * <p>
 * 1. add `"local:{}` to "files" node in layotto/configs/config_file.json
 * 2. start server by `./layotto start -c ../../configs/config_file.json`
 */
public class File {

    private static final Logger logger    = LoggerFactory.getLogger(File.class.getName());

    static String               storeName = "local";
    static String               fileName  = "/tmp/test.log";

    public static void main(String[] args) throws Exception {

        RuntimeClient client = new RuntimeClientBuilder()
            .withPort(RuntimeProperties.DEFAULT_PORT)
            .build();

        putFile(client);
        getFile(client);
        listFile(client);
        getFileMeta(client);
        delFile(client);
    }

    public static void putFile(RuntimeClient client) throws Exception {

        PutFileRequest request = new PutFileRequest();
        request.setStoreName(storeName);
        request.setFileName(fileName);

        Map<String, String> meta = new HashMap<>();
        meta.put("FileMode", "521");
        meta.put("FileFlag", "0777");
        request.setMetaData(meta);

        request.setIn(new ByteArrayInputStream("hello world".getBytes()));

        client.putFile(request, 3000);
    }

    public static void getFile(RuntimeClient client) throws Exception {

        GetFileRequest request = new GetFileRequest();
        request.setStoreName(storeName);
        request.setFileName(fileName);

        Map<String, String> meta = new HashMap<>();
        meta.put("k1", "v1");
        request.setMetaData(meta);

        GetFileResponse resp = client.getFile(request, 3000);

        InputStream reader = resp.getIn();

        byte[] buf = new byte[128];
        for (int len = reader.read(buf); len > 0; len = reader.read(buf)) {
            logger.info(new String(buf, 0, len));
        }
    }

    public static void delFile(RuntimeClient client) throws Exception {

        DelFileRequest request = new DelFileRequest();
        request.setStoreName(storeName);
        request.setFileName(fileName);

        client.delFile(request, 3000);
    }

    public static void listFile(RuntimeClient client) throws Exception {

        ListFileRequest request = new ListFileRequest();
        request.setStoreName(storeName);
        request.setMarker("test.log");
        request.setName("/tmp");
        request.setPageSize(10);

        ListFileResponse resp = client.listFile(request, 3000);

        for (FileInfo f : resp.getFiles()) {
            logger.info(f.getFileName());
        }
    }

    public static void getFileMeta(RuntimeClient client) throws Exception {

        GetMetaRequest request = new GetMetaRequest();
        request.setStoreName(storeName);
        request.setFileName(fileName);

        GetMeteResponse response = client.getFileMeta(request, 3000);
        logger.info(response.getLastModified());
        logger.info("" + response.getMeta().size());
    }
}
